### ISSUE TITLE
can't evaluate field targetValue in type interface

### ISSUE THREAD

**Issue by ramili4** [helm/helm#13394]
Hi there! Trying to implement hpa with custom metrics from prometheus adapter, but run into this error

- fatal: [localhost]: FAILED! => {"changed": false, "command": "/usr/local/bin/helm upgrade -i --reset-values --wait --timeout 5m --atomic --values=../values/service/app-name/dev@dev/values.yaml --values=../values/all/hpa_values.yaml -f=/tmp/tmpbir4ghm1.yml app ../helm-chart\", "msg": "Failure when executing Helm command. Exited 1.\nstdout: \nstderr: coalesce.go:223: warning: destination for common-chart.hpa.enabled is a table. Ignoring non-table value (false)\nError: UPGRADE FAILED: template: common-chart/templates/hpa.yaml:26:25: executing \"my-common-chart/templates/hpa.yaml\" at <.Values.hpa_metrics.targetValue>: can't evaluate field targetValue in type interface {}\n", "stderr": "coalesce.go:223: warning: destination for my-common-chart.hpa.enabled is a table. Ignoring non-table value (false)\nError: UPGRADE FAILED: template: my-common-chart/templates/hpa.yaml:26:25: executing \"my-common-chart/templates/hpa.yaml\" at <.Values.hpa_metrics.targetValue>: can't evaluate field targetValue in type interface {}\n", "stderr_lines": ["coalesce.go:223: warning: destination for my-common-chart.hpa.enabled is a table. Ignoring non-table value (false)", "Error: UPGRADE FAILED: template: my-common-chart/templates/hpa.yaml:26:25: executing \"my-common-chart/templates/hpa.yaml\" at <.Values.hpa_metrics.targetValue>: can't evaluate field targetValue in type interface {}"], "stdout": "", "stdout_lines":

Here's the content of hpa_values.yaml:
hpa: 
  enabled: true
  minReplicas: 2
  maxReplicas: 8
  
hpa_metrics:
  targetValue: 2

hpa_metrics:
  - type: Object
    object:
      metric:
        name: container_cpu_usage
      describedObject:
        apiVersion: apps/v1
        kind: Deployment
        name: "{{ app_name }}"
      target:
        type: Value
        value: 2

And this is the HPA template:
{{- $targetValue := .Values.hpa_metrics.targetValue | default 2 -}}
{{- if .Values.hpa }}
{{- if .Values.hpa.enabled }}
apiVersion: autoscaling/v2
kind: HorizontalPodAutoscaler
metadata:
  name: {{ .Values.fullnameOverride }}-hpa
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: {{ .Values.fullname}}
  minReplicas: {{ .Values.hpa.minReplicas | default 2 }}
  maxReplicas: {{ .Values.hpa.maxReplicas | default 10 }}
  metrics:
  - type: Object
    object:
      metric:
        name: container_cpu_usage
      describedObject:
        apiVersion: apps/v1
        kind: Deployment
        name: {{ .Values.fullname }}
      target:
        type: Value
        value: {{ $targetValue }}
{{- end }}
{{- end }}
What can be wrong here? I suspect container_cpu_usage metric might not be available or somehow misconfigured in Prometheus adapter.

**Comment by banjoh** [comment]
Its hard to tell whether this is related to helm not working as should. Do you have a reproducer helm chart and a set of values I can can run? I'm looking for enough information to run `helm template <chart> <values.yaml>`

**Comment by ramili4** [comment]
nope, I don't have it-( It has a bunch of values files and a corresponding playbook that deploys services

**Comment by banjoh** [comment]
I see. Are you able to provide the exact helm values file that you pasted above? From the error output, it looks like your problem is around `Values.hpa_metrics.targetValue`. I can not see such a field in https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/horizontal-pod-autoscaler-v2/. Check your configuration against this spec.

This is my best guess.

**Comment by ramili4** [comment]
That's the custom metric from Prometheus adapter. I guess I might have misplaced some values in a variety of values files. I'm fuming already, but the issue is on the surface, I just don't see it

**Comment by ramili4** [comment]
Can I reference the entire hpa_metrics section in my values.yaml file with this command - {% for metric in hpa_metrics %}
    - {{ metric }}
{% endfor %}? 

hpa_metrics section
hpa_metrics:
  - type: Object
    object:
      metric:
        name: container_cpu_usage
      describedObject:
        apiVersion: apps/v1
        kind: Deployment
        name: "{{ application_name }}"
      target:
        type: Value
        value: 2
